cmd/supernode/app/root.go

```diff
@@ -65,121 +121,211 @@ func setupFlags(cmd *cobra.Command, opt *Options) {
-		return fmt.Errorf("failed to create home dir %s: %v", options.HomeDir, err)
+// bindRootFlags binds flags on rootCmd to the given viper instance.
+func bindRootFlags(v *viper.Viper) error {
+	flags := []struct {
+		key  string
+		flag string
+	}{
+		{
+			key:  "config",
+			flag: "config",
+		},
+		{
+			key:  "base.listenPort",
+			flag: "port",
+		},
+		{
+			key:  "base.downloadPort",
+			flag: "download-port",
+		},
+		{
+			key:  "base.homeDir",
+			flag: "home-dir",
+		},
+		{
+			key:  "base.downloadPath",
+			flag: "download-path",
+		},
+		{
+			key:  "base.systemReservedBandwidth",
+			flag: "system-bandwidth",
+		},
+		{
+			key:  "base.maxBandwidth",
+			flag: "max-bandwidth",
+		},
+		{
+			key:  "base.schedulerCorePoolSize",
+			flag: "pool-size",
+		},
+		{
+			key:  "base.enableProfiler",
+			flag: "profiler",
+		},
+		{
+			key:  "base.debug",
+			flag: "debug",
+		},
+		{
+			key:  "base.peerUpLimit",
+			flag: "up-limit",
+		},
+		{
+			key:  "base.peerDownLimit",
+			flag: "down-limit",
+		},
+		{
+			key:  "base.advertiseIP",
+			flag: "advertise-ip",
+		},
+		{
+			key:  "base.failAccessInterval",
+			flag: "fail-access-interval",
+		},
+		{
+			key:  "base.gcInitialDelay",
+			flag: "gc-initial-delay",
+		},
+		{
+			key:  "base.gcMetaInterval",
+			flag: "gc-meta-interval",
+		},
+		{
+			key:  "base.taskExpireTime",
+			flag: "task-expire-time",
+		},
+		{
+			key:  "base.peerGCDelay",
+			flag: "peer-gc-delay",
+		},
 	}
 
-	// initialize supernode logger.
-	if err := initLog(logrus.StandardLogger(), "app.log"); err != nil {
-		return err
+	for _, f := range flags {
+		if err := v.BindPFlag(f.key, rootCmd.Flag(f.flag)); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	// initialize dfget logger.
-	dfgetLogger := logrus.New()
-	if err := initLog(dfgetLogger, "dfget.log"); err != nil {
+// readConfigFile reads config file into the given viper instance. If we're
+// reading the default configuration file and the file does not exist, nil will
+// be returned.
+func readConfigFile(v *viper.Viper, cmd *cobra.Command) error {
+	v.SetConfigFile(v.GetString("config"))
+	v.SetConfigType("yaml")
+
+	if err := v.ReadInConfig(); err != nil {
+		// when the default config file is not found, ignore the error
+		if os.IsNotExist(err) && !cmd.Flag("config").Changed {
+			return nil
+		}
 		return err
 	}
 
-	if err := initConfig(); err != nil {
-		return err
-	}
+	return nil
+}
 
-	// set supernode advertise ip
-	if stringutils.IsEmptyStr(cfg.AdvertiseIP) {
-		if err := setAdvertiseIP(); err != nil {
-			return err
-		}
+// getConfigFromViper returns supernode config from the given viper instance
+func getConfigFromViper(cmd *cobra.Command, v *viper.Viper) (*config.Config, error) {
+	cfg := config.NewConfig()
+
+	if err := v.Unmarshal(cfg, func(dc *mapstructure.DecoderConfig) {
+		dc.TagName = "yaml"
+		dc.DecodeHook = decodeWithYAML(
+			reflect.TypeOf(time.Second),
+			reflect.TypeOf(rate.B),
```

Comment:
Since #878 has been merged, maybe you should make a handle for type `fileutils.Fsize`.